src/lib.rs

```diff
@@ -0,0 +1,386 @@
+extern crate libc;
+
+mod ffi;
+
+use std::path::Path;
+
+pub type Result<T> = std::result::Result<T, String>;
+
+pub type Ts = ffi::rill_ts_t;
+pub type Key = ffi::rill_key_t;
+pub type Val = ffi::rill_val_t;
+pub type KV = ffi::rill_kv;
+
+// This function is kind-of insane. I really hope there's an easier
+// way to do this...
+fn err<T>() -> Result<T> {
+    let err = unsafe{ ffi::rill_errno_thread() };
+
+    let mut buf: [u8; 1024] = [0; 1024];
+    let len = unsafe{
+        let ptr = buf.as_mut_ptr() as *mut libc::c_char;
+        let len = ffi::rill_strerror(err, ptr, buf.len());
+        if len >= buf.len() { buf.len() } else { len + 1 }
+    };
+
+    match std::ffi::CStr::from_bytes_with_nul(&buf[0..len]) {
+        Err(err) => Err(format!("invalid error string: {}", err)),
+        Ok(msg) => match msg.to_str() {
+            Err(err) => Err(format!("invalid error string: {}", err)),
+            Ok(msg) =>Err(String::from(msg)),
+        }
+    }
+}
+
+// \todo Need to add iterator support.
+pub struct Pairs { pairs: *mut ffi::rill_pairs }
+
+impl Pairs {
+    pub fn with_capacity(cap: usize) -> Result<Pairs> {
+        let ptr = unsafe { ffi::rill_pairs_new(cap) };
+        if !ptr.is_null() { Ok(Pairs{ pairs: ptr }) } else { err() }
+    }
+
+    pub fn clear(&mut self) {
+        unsafe { ffi::rill_pairs_clear(self.pairs) }
+    }
+
+    pub fn capacity(&self) -> usize {
+        unsafe { ffi::rill_pairs_cap(self.pairs) }
+    }
+
+    pub fn len(&self) -> usize {
+        unsafe { ffi::rill_pairs_len(self.pairs) }
+    }
+
+    pub fn get(&self, index: usize) -> &KV {
+        unsafe { &*ffi::rill_pairs_get(self.pairs, index) }
+    }
+
+    pub fn push(&mut self, key: Key, val: Val) -> Result<()> {
+        let result = unsafe { ffi::rill_pairs_push(self.pairs, key, val) };
+        if result.is_null() { return err() };
+        self.pairs = result;
+        Ok(())
+    }
+
+    pub fn compact(&mut self) {
+        unsafe { ffi::rill_pairs_compact(self.pairs) }
+    }
+}
+
+impl Drop for Pairs {
+    fn drop(&mut self) {
+        unsafe { ffi::rill_pairs_free(self.pairs) }
+    }
+}
+
+#[test]
```

Comment:
Please follow the [book](https://doc.rust-lang.org/stable/book/second-edition/ch11-01-writing-tests.html), i.e. having a `#[cfg(test)]` attribute on a `mod tests` to avoid having these functions in the final binary.